src/Microsoft.AspNetCore.Blazor.Browser/Interop/ArgumentList.cs

```diff
@@ -0,0 +1,238 @@
+﻿using System;
+using System.Linq;
+using System.Reflection;
+
+namespace Microsoft.AspNetCore.Blazor.Browser.Interop
+{
+    internal class ArgumentList
```

Comment:
Although the `ArgumentsList` approach looks pretty complete here, it does come with quite a lot of sophisticatation (e.g., having to pack args into a nested structure), and some runtime overhead (e.g., a fair chunk of reflection and allocating lots of arrays).

I have an idea for a different way of passing and JSON-deserializing the args that avoids the need for this altogether. However I don't want to derail you at this stage! So,

* If you are interested in changing to a different approach at this stage, let me know and we can discuss ideas
* If you prefer to keep `ArgumentsList` and the related machinery now, would you be OK with us possibly changing it later?

Reply:
I didn't want to have to deal with my own parsing or fiddle with the parser. If this was JSON.NET, I would have just read the types myself using their lower level parsing API.

In most cases this will allocate one array I believe + 1 type for the packing, and we can make that a struct to avoid the allocation if we want.

Realistically, most API calls have less than 8 parameters, so I'm not sure a more sophisticated approach is worth here.

That said, I'm not opposed to changing this later if we can prove the we can get a significant speed up out of it.

Reply:
The main benefit I'd see would be not maintaining a lot of code and tests that's specific to different numbers of args. But let's not change it for this PR - we can always add it later. The `ArgumentsList` approach is totally fine for now.